File: bugsnag-android-core/src/main/java/com/bugsnag/android/Thread.java
package com.bugsnag.android;

import androidx.annotation.NonNull;

import java.io.IOException;

@SuppressWarnings("ConstantConditions")
public class Thread implements JsonStream.Streamable {

    private final ThreadImpl impl;
    private final Logger logger;

    Thread(
            long id,
            @NonNull String name,
            @NonNull ThreadType type,
            boolean errorReportingThread,
            @NonNull Stacktrace stacktrace,
            @NonNull Logger logger) {
        this.impl = new ThreadImpl(id, name, type, errorReportingThread, stacktrace);
        this.logger = logger;
    }

    private void error(String property) {
        logger.e("Invalid null value supplied to thread." + property + ", ignoring");
    }

    public void setId(long id) {
        impl.setId(id);
    }

    public long getId() {
        return impl.getId();
    }

    public void setName(@NonNull String name) {
        if (name != null) {
            impl.setName(name);
        } else {
            error("name");
        }
    }

    @NonNull
    public String getName() {
        return impl.getName();
    }

    public void setType(@NonNull ThreadType type) {
        if (type != null) {
            impl.setType(type);
        } else {
            error("type");
        }
    }

    @NonNull
    public ThreadType getType() {
        return impl.getType();
    }

    public void setErrorReportingThread(boolean errorReportingThread) {
        impl.setErrorReportingThread(errorReportingThread);
    }

    public boolean getErrorReportingThread() {
        return impl.isErrorReportingThread();
    }

    @Override
    public void toStream(@NonNull JsonStream stream) throws IOException {
        impl.toStream(stream);
    }
}

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Thread.kt
package com.bugsnag.android

import java.io.IOException

/**
 * A representation of a thread recorded in an [Event]
 */
class Thread internal constructor(

    /**
     * The unique ID of the thread (from [java.lang.Thread])
     */
    val id: Long,

    /**
     * The name of the thread (from [java.lang.Thread])
     */
    val name: String,

    /**
     * The type of thread based on the originating platform (intended for internal use only)
     */
    val type: Type,

    /**
     * Whether the thread was the thread that caused the event
     */
    val isErrorReportingThread: Boolean,
    stacktrace: Stacktrace
) : JsonStream.Streamable {

    /**
     * Represents the type of thread captured
     */
    enum class Type(internal val desc: String) {

        /**
         * A thread captured from Android's JVM layer
         */
        ANDROID("android"),

        /**
         * A thread captured from JavaScript
         */
        BROWSER_JS("browserjs")
    }

    /**
     * Controls whether we should capture and serialize the state of all threads at the time
     * of an error.
     */
    enum class ThreadSendPolicy {

        /**
         * Threads should be captured for all events.
         */
        ALWAYS,

        /**
         * Threads should be captured for unhandled events only.
         */
        UNHANDLED_ONLY,

        /**
         * Threads should never be captured.
         */
        NEVER;

        internal companion object {
            fun fromString(str: String) = values().find { it.name == str } ?: ALWAYS
        }
    }

    /**
     * A representation of the thread's stacktrace
     */
    var stacktrace: MutableList<Stackframe> = stacktrace.trace.toMutableList()

    @Throws(IOException::class)
    override fun toStream(writer: JsonStream) {
        writer.beginObject()
        writer.name("id").value(id)
        writer.name("name").value(name)
        writer.name("type").value(type.desc)

        writer.name("stacktrace")
        writer.beginArray()
        stacktrace.forEach { writer.value(it) }
        writer.endArray()

        if (isErrorReportingThread) {
            writer.name("errorReportingThread").value(true)
        }
        writer.endObject()
    }
}
